metricflow/inference/solver/weighted_tree.py

```diff
@@ -0,0 +1,163 @@
+                weights=weights,
+                non_complimentary_weights=non_complimentary_weights,
+                signals_by_type=signals_by_type,
+            )
+
+        weights[root] = sum(self._weighter_function(signal.confidence) for signal in signals_by_type[root])
+        non_complimentary_weights[root] = sum(
+            self._weighter_function(signal.confidence)
+            for signal in signals_by_type[root]
+            if not signal.is_complimentary
+        )
+
+        weights[root] += sum(non_complimentary_weights[child] for child in root.children)
+
+        return weights
+
+    def _get_cumulative_weights(self, signals: List[InferenceSignal]) -> Dict[InferenceSignalNode, int]:
+        """Get the cumulative weights dict for a list of signals"""
+        signals_by_type: Dict[InferenceSignalNode, List[InferenceSignal]] = defaultdict(list)
+        for signal in signals:
+            signals_by_type[signal.type_node].append(signal)
+
+        return self._get_cumulative_weights_for_root(
+            root=InferenceSignalType.UNKNOWN,
+            weights=defaultdict(lambda: 0),
+            non_complimentary_weights=defaultdict(lambda: 0),
+            signals_by_type=signals_by_type,
+        )
+
+    def solve_column(self, column: SqlColumn, signals: List[InferenceSignal]) -> InferenceResult:
+        """Find the appropriate type for a column by traversing through the type tree.
+
+        It traverses the tree by giving weights to all nodes and greedily finding the path with the most
+        weight until it either finds a leaf or there is a "weight bifurcation" in the path with respect
+        to the provided `weight_percent_threshold`.
+        """
+        if len(signals) == 0:
+            return InferenceResult(
+                column=column,
+                type_node=InferenceSignalType.UNKNOWN,
+                reasons=[],
+                problems=[
+                    "No signals were extracted for this column",
+                    "Inference solver could not determine a type for this column",
+                ],
+            )
+
+        reasons_by_type: Dict[InferenceSignalNode, List[str]] = defaultdict(list)
+        for signal in signals:
+            reasons_by_type[signal.type_node].append(f"{signal.reason} ({signal.type_node.name})")
+
+        node_weights = self._get_cumulative_weights(signals)
+
+        reasons: List[str] = []
+        problems: List[str] = []
+        node = InferenceSignalType.UNKNOWN
+        while len(node.children) > 0:
+            children_weight_total = sum(node_weights[child] for child in node.children)
+
+            if children_weight_total == 0:
+                break
+
+            next_node = None
+            for child in node.children:
+                if node_weights[child] / children_weight_total >= self._weight_percent_threshold:
+                    next_node = child
+                    reasons += reasons_by_type[child]
+                    break
```

Comment:
Careful - first one in wins, but with `self._weight_percent_threshold` in [0.5, 1] you can end up with 2 equally good matches. See above for the simplest, most robust fix - I'm not recommending you do it here because your chosen weights can lead to weirdness on things like thresholds of 0.6 or 0.8 or 0.75.